src/pymor/operators/block.py

```diff
@@ -225,10 +225,22 @@ def assemble(self, mu=None):
             return self.__class__(blocks)
 
     def assemble_lincomb(self, operators, coefficients, solver_options=None, name=None):
+        assert operators[0] is self
+
+        # return ShiftedSecondOrderOperator if possible
+        if (len(operators) == 2 and self.num_source_blocks == 2 and self.num_range_blocks == 2 and
+                isinstance(self._blocks[0, 0], IdentityOperator) and isinstance(operators[1], SecondOrderOperator)):
+            return ShiftedSecondOrderOperator(self._blocks[1, 1],
+                                              operators[1].D,
+                                              operators[1].K,
+                                              coefficients[0],
+                                              coefficients[1])
```

Comment:
Hmm, I don't quite like this special case here. If would prefer something like
```
if len(operators) == 2 and isinstance(operators[1], SecondOrderOperator):
    return operators[1].assemble_lincomb(operators[::-1], coefficients[::-1], solver_options, name)
```

In general it seems to me that this is another case where inheritance doesn't quite fit our needs. For the not to near future we should consider moving most of the functionality into an `assemble `RuleTable`. `assemble_lincomb` was never really inteded for such use cases. It should mostly be concerned with forming linear combinations of backend matrices.

Reply:
This might be a more general issue of solving shifted linear systems. @sdrave, would a PDE solver know how to solve a linear systems with a `LincombOperator` of the stiffness and inner product matrices?

Reply:
What are you thinking of? A special preconditioner?

Reply:
I mean, would `lgmres` be avoided when solving with s*E-A, where E and A come from a PDE discretization?

Reply:
Do you know of any PDE solver library which provides a method for a shifted linear system solve? I would expect this, since it appears, e.g., in the implicit Euler method.
Another option might be to define a new PDE, with an additional term, which would give the shifted linear system as the stiffness matrix when discretized.

Reply:
I am not sure if I understand you correctly.

To give an example: when using pyMORs `ImplicitEulerTimeStepper` with a discretization coming from our PDE solver library [`dune-gdt`](https://github.com/dune-community/dune-gdt), what happens for instance in [`implicit_euler`](https://github.com/pymor/pymor/blob/master/src/pymor/algorithms/timestepping.py#L162), is that we obtain a `LincombOperator` holding the (wrapped) PDE solvers matrices. When solving the system in [`implicit_euler`](https://github.com/pymor/pymor/blob/master/src/pymor/algorithms/timestepping.py#L177) with this new operator, [`assemble_lincomb`](https://github.com/pymor/pymor/blob/master/src/pymor/operators/constructions.py#L126) is called on the wrapped matrix of the PDE solver (which is why we pass on the `solver_options`). The implementation of `assemble_lincomb` of the wrapped `dune-gdt` matrices calls something like `axpy` on the PDE solvers matrices to create a new matrix with appropriate sparsity pattern. This is PDE solver bindings specific code, but adding PDE solver bindings without an efficient `assemble_lincomb` implementation for the PDE solvers linear algebra backend would be far for optimal, to put it diplomatically :). Inverting this matrix again then calls the linear solver of the underlying PDE solver.

Thus, in this specific case the PDE solver does not need to know that he is used in the context of a shifted system (let alone in the context of an instationary problem). In addition, `axpy` of two sparse matrices is usually much cheaper than anything else we do in a PDE solver (walking the grid and evaluating shape functions is mush more costly than blas operations).

If you think of matrix-free PDE solvers, one would use an iterative solver where in each iteration you would need to walk the grid and apply the local action of the associated mass operator. So here the PDE solver would known about the shifted system, but without any real benefit for us.